src/vsr/replica.zig

```diff
@@ -1085,14 +1099,23 @@ pub fn Replica(
             assert(!self.do_view_change_quorum);
             self.do_view_change_quorum = true;
 
+            // The `prepare_timestamp` prevents a primary's own clock from running backwards.
+            // A primary must not timestamp a prepare before another prepare it has prepared.
+            // Therefore, the `prepare_timestamp`:
+            // 1. is advanced if behind the cluster, but never reset if ahead of the cluster, i.e.
+            // 2. may not always reflect the timestamp of the latest prepared op, and
+            // 3. should be advanced before discarding the timestamps of any uncommitted headers.
+            if (self.state_machine.prepare_timestamp < latest.timestamp) {
+                self.state_machine.prepare_timestamp = latest.timestamp;
+            }
+
+            // TODO At present, discard_uncommitted_headers() does not consider interactions with
+            // the hash chain, i.e. that could lead to a hash chain swing, when it removes
+            // uncommitted ops. This is a correctness bug, where old ops may not be replaced in the
+            // view change. Rather, discard_uncommitted_headers() must fix up self.op after discard.
```

Comment:
> that could lead to a hash chain swing

I'm not sure I understand what a "hash chain swing" is.

I understand that the new `self.journal.set_header_as_dirty` call above can replace an op, potentially breaking the hash chain between `commit_min` and `commit_min + 1` (for every `do_view_change`'s `commit_min`). So we may need to discard some ops that follow a `commit_min`.

But the comment implies the "swing" hasn't happened yet?

Reply:
A “hash chain swing” or “quorum swing” would be where a prepare is replaced with another, undermining the guarantee to the previous primary, meaning that the commit quorum “swings” to a new decision by another primary, forking the log if the prepare was in fact committed.